### PR TITLE
Refactors weather into a subsystem

### PR DESCRIPTION
**Pull request by ghost**
Weather has been refactored from a weather control machine on the lavaland z-level into a subsystem. All existing weather has been changed to accommodate this change, and their code has been optimized by using `addtimer()` instead of `sleep()`.

The new subsystem also supports adding weather to any z-level; for instance, if you made a weather called `/datum/weather/rain_storm` and made it target z-level 3 with a 100% probability, the rain storm would occur as much as possible with 5-10 minute intermissions. These intermissions take into account the weather's duration.

### CODE REVIEW

File: code/controllers/subsystem/weather.dm [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+	var/list/eligible_zlevels = list(ZLEVEL_LAVALAND)
+
+/datum/subsystem/weather/New()
+	NEW_SS_GLOBAL(SSweather)
+	for(var/V in subtypesof(/datum/weather))
+		var/datum/weather/W = V
+		existing_weather += new W
+
+/datum/subsystem/weather/fire()
+	for(var/datum/weather/W in processing)
```
Comment (RemieRichards) [inline code/controllers/subsystem/weather.dm]: this should be an istypeless forloop.

File: code/controllers/subsystem/weather.dm [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+		if(W.aesthetic)
+			continue
+		for(var/mob/living/L in mob_list)
+			var/area/A = get_area(L)
+			if(L.z == W.target_z && L.weather_immunities & !W.immunity_type && A in W.impacted_areas)
+				W.impact(L)
+	for(var/Z in eligible_zlevels)
+		var/list/possible_weather_for_this_z = list()
+		for(var/datum/weather/WE in existing_weather)
```
Comment (RemieRichards) [inline code/controllers/subsystem/weather.dm]: this should be an istypeless forloop.

File: code/controllers/subsystem/weather.dm [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+				possible_weather_for_this_z[WE] = WE.probability
+		var/datum/weather/W = pickweight(possible_weather_for_this_z)
+		run_weather(W.name)
+		eligible_zlevels -= Z
+		addtimer(src, "make_z_eligible", rand(3000, 6000) + W.weather_duration_upper, Z) //Around 5-10 minutes between weathers
+
+/datum/subsystem/weather/proc/run_weather(weather_name)
+	if(!weather_name)
+		return
+	for(var/datum/weather/W in existing_weather)
```
Comment (RemieRichards) [inline code/controllers/subsystem/weather.dm]: this should be an istypeless forloop.

File: code/datums/weather/weather.dm [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+
+/datum/weather/Destroy()
+	SSweather.existing_weather -= src
+	..()
+
+/datum/weather/proc/telegraph()
+	if(stage == STARTUP_STAGE)
+		return
+	stage = STARTUP_STAGE
+	for(var/area/A in get_areas(area_type))
```
Comment (RemieRichards) [inline code/datums/weather/weather.dm]: this should be an istypeless forloop.

File: code/datums/weather/weather.dm [hunk 5]
```diff
@@ -0,0 +1,120 @@
...
+		if(A.z == target_z)
+			impacted_areas |= A
+	weather_duration = rand(weather_duration_lower, weather_duration_upper)
+	update_areas()
+	for(var/mob/M in player_list)
```
Comment (RemieRichards) [inline code/datums/weather/weather.dm]: this should be an istypeless forloop.

File: code/datums/weather/weather.dm [hunk 6]
```diff
@@ -0,0 +1,120 @@
...
+			if(telegraph_sound)
+				M << sound(telegraph_sound)
+	addtimer(src, "start", telegraph_duration)
+
+/datum/weather/proc/start()
+	if(stage >= MAIN_STAGE)
+		return
+	stage = MAIN_STAGE
+	update_areas()
+	for(var/mob/M in player_list)
```
Comment (RemieRichards) [inline code/datums/weather/weather.dm]: this should be an istypeless forloop.

File: code/datums/weather/weather.dm [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+				M << sound(weather_sound)
+	SSweather.processing |= src
+	addtimer(src, "wind_down", weather_duration)
+
+/datum/weather/proc/wind_down()
+	if(stage >= WIND_DOWN_STAGE)
+		return
+	stage = WIND_DOWN_STAGE
+	update_areas()
+	for(var/mob/M in player_list)
```
Comment (RemieRichards) [inline code/datums/weather/weather.dm]: this should be an istypeless forloop.

File: code/datums/weather/weather.dm [hunk 8]
```diff
@@ -0,0 +1,120 @@
...
+	if(stage == END_STAGE)
+		return
+	stage = END_STAGE
+	update_areas()
+
+/datum/weather/proc/impact(mob/living/L) //What effect does this weather have on the hapless mob?
+	return
+
+/datum/weather/proc/update_areas()
+	for(var/area/N in impacted_areas)
```
Comment (RemieRichards) [inline code/datums/weather/weather.dm]: this should be an istypeless forloop.

File: code/datums/weather/weather_types.dm [hunk 9]
```diff
@@ -0,0 +1,112 @@
...
+	weather_duration_upper = 300
+
+	end_message = "<span class='danger'>At last, the darkness recedes.</span>"
+	end_duration = 0
+
+	area_type = /area
+	target_z = ZLEVEL_STATION
+
+/datum/weather/advanced_darkness/update_areas()
+	for(var/area/A in impacted_areas)
```
Comment (RemieRichards) [inline code/datums/weather/weather_types.dm]: this should be an istypeless forloop.

File: code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm [hunk 10]
```diff
@@ -118,37 +118,37 @@
...
+		user << "<span class='warning'>You can't seem to control the weather here!</span>"
+		return
 
-	if(!linked_machine || linked_machine.z != user.z)
-		for(var/obj/machinery/lavaland_controller/controller in machines)
-			if(controller.z == user.z)
-				linked_machine = controller
-				break
+	var/datum/weather/ash_storm/A
+	for(var/datum/weather/W in SSweather.existing_weather)
```
Comment (RemieRichards) [inline code/modules/mob/living/simple_animal/hostile/megafauna/legion.dm]: this should be an istypeless forloop.

File: code/modules/ruins/lavaland_ruin_code.dm [hunk 11]
```diff
@@ -1,36 +1,5 @@
...
-	ongoing_weather = null
-
-/obj/machinery/lavaland_controller/Destroy(force)
-	if(force)
-		. = ..()
-	else
-		return QDEL_HINT_LETMELIVE
-
-
-/obj/structure/fans/tiny/invisible //For blocking air in ruin doorways
```
Comment (optimumtact) [inline code/modules/ruins/lavaland_ruin_code.dm]: Did you mean to remove this?

Comment (coiax) [inline code/modules/ruins/lavaland_ruin_code.dm]: TINY FAN

**Comment by Xhuis** [comment]
Tiny fan re-added and weather controller removed from lavaland. Didn't use the map merger because it seems to have issues with `lavaland.dmm`.

**Comment by optimumtact** [comment]
what kind of issues

**Comment by Xhuis** [comment]
You can see it in the other commit. It replaces a single tile that has two medkits _every time_ and it doesn't seem to save removals.

**Comment by coiax** [comment]
@xxalpha @xxalpha @xxalpha 

Say their name three times to summon them.

**Comment by xxalpha** [comment]
Mmmiracles edited lavaland just fine.

**Comment by Xhuis** [comment]
Guess it's just me, then. The TGM mapmerger refuses to work for me with lavaland.

**Comment by Xhuis** [comment]
Closing this until I can figure out why the mapmerger won't work for me.

**Comment by Xhuis** [comment]
It seems like other people are having the same issue. For the time being, I'm just going to remove the functionality of the weather controller. New commit coming soon.

**Comment by Xhuis** [comment]
* Replaced a bunch of for loops with istypeless for loops.
* Removed map edits and removed the functionality of the weather machine pending its removal.
* Re-added invisible tiny fans.

**Comment by Xhuis** [comment]
This is no longer WIP and should be merge-ready.

File: code/modules/admin/secrets.dm [hunk 12]
```diff
@@ -427,8 +427,7 @@
 				L.fix()
 
 		if("floorlava")
-			var/datum/weather/floor_is_lava/storm = new /datum/weather/floor_is_lava
-			storm.weather_start_up()
+			SSweather.run_weather("the floor is lava")
```
Comment (KorPhaeron) [inline code/modules/admin/secrets.dm]: why would you merge these magic string memes @optimumtact

File: code/controllers/subsystem/weather.dm [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+		eligible_zlevels -= Z
+		addtimer(src, "make_z_eligible", rand(3000, 6000) + W.weather_duration_upper, Z) //Around 5-10 minutes between weathers
+
+/datum/subsystem/weather/Initialize(start_timeofday)
+	..()
+	for(var/V in subtypesof(/datum/weather))
+		var/datum/weather/W = V
+		existing_weather |= new W
+
+/datum/subsystem/weather/proc/run_weather(weather_name)
```
Comment (KorPhaeron) [inline code/controllers/subsystem/weather.dm]: @JJRcop look at this aaaaaaaa

File: code/modules/admin/secrets.dm (hunk 12 above)
Comment (KorPhaeron) [inline code/modules/admin/secrets.dm]: @JJRcop

Comment (Xhuis) [inline code/modules/admin/secrets.dm]: Should I be using defines? An ID var in weather itself?

Comment (xxalpha) [inline code/modules/admin/secrets.dm]: Yes, number values with descriptive defines. There is no reason to use strings for this sort of stuff.

**Pull request merged**